enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/replication/RaftReplicator.java

```diff
@@ -72,53 +72,76 @@ public RaftReplicator( LeaderLocator leaderLocator, MemberId me, Outbound<Member
+            }
+            catch ( InterruptedException e )
+            {
+                throw new ReplicationFailureException( "Interrupted while waiting for replication credits", e );
+            }
         }
         else
         {
-            return replicate0( command, trackResult );
+            return replicate0( command, trackResult, originalLeader );
         }
     }
 
-    private Future<Object> replicate0( ReplicatedContent command, boolean trackResult ) throws InterruptedException
+    private Future<Object> replicate0( ReplicatedContent command, boolean trackResult, MemberId leader ) throws ReplicationFailureException
     {
+        assertNoLeaderSwitch( leader );
+
         OperationContext session = sessionPool.acquireSession();
 
         DistributedOperation operation = new DistributedOperation( command, session.globalSession(), session.localOperationId() );
         Progress progress = progressTracker.start( operation );
 
         TimeoutStrategy.Timeout progressTimeout = progressTimeoutStrategy.newTimeout();
         TimeoutStrategy.Timeout leaderTimeout = leaderTimeoutStrategy.newTimeout();
-        do
+        try
         {
-            assertDatabaseNotShutdown();
-            try
+            do
             {
-                // blocking at least until the send has succeeded or failed before retrying
-                outbound.send( leaderLocator.getLeader(), new RaftMessages.NewEntry.Request( me, operation ), true );
+                assertDatabaseNotShutdown();
+                try
+                {
+                    // blocking at least until the send has succeeded or failed before retrying
+                    outbound.send( leader, new RaftMessages.NewEntry.Request( me, operation ), true );
 
-                leaderTimeout = leaderTimeoutStrategy.newTimeout();
+                    leaderTimeout = leaderTimeoutStrategy.newTimeout();
 
-                progress.awaitReplication( progressTimeout.getMillis() );
-                progressTimeout.increment();
-            }
-            catch ( InterruptedException e )
-            {
-                progressTracker.abort( operation );
-                throw e;
-            }
-            catch ( NoLeaderFoundException e )
-            {
-                log.debug( "Could not replicate operation " + operation + " because no leader was found. Retrying.", e );
-                Thread.sleep( leaderTimeout.getMillis() );
-                leaderTimeout.increment();
+                    progress.awaitReplication( progressTimeout.getMillis() );
+                    progressTimeout.increment();
+                    leader = leaderLocator.getLeader();
+                }
+                catch ( NoLeaderFoundException e )
+                {
+                    log.debug( "Could not replicate operation " + operation + " because no leader was found. Retrying.", e );
+                    Thread.sleep( leaderTimeout.getMillis() );
```

Comment:
A bit confused about this part, as it looks like you're creating a new leader timeout inside the loop. Also, is the `increment` only for testing? The ConstantTimeout which gets injected for the leaderStrategy in prod has a noop implementation for it.

Reply:
@hugofirth This would be clearer if a backoff timeout were injected. `leaderTimeoutStrategy.newTimeout()` would create a timeout (after a success), and the `increment` happen on failure. So if `outbound.send` kept failing the timeout would keep being incremented; as soon as it passes the timeout is reset.

Reply:
The retry strategy was specifically for `leaderLocator.getLeader()`, but as that is no longer in the block is the retry still relevant?